cluster/distributedtask/scheduler.go

```diff
@@ -0,0 +1,365 @@
+					"taskVersion": taskDesc.Version,
+					"error":       err,
+				}).Error("failed to clean up local distributed task state")
+				continue
+			}
+
+			s.logger.WithFields(logrus.Fields{
+				"namespace":   namespace,
+				"taskID":      taskDesc.ID,
+				"taskVersion": taskDesc.Version,
+			}).Info("cleaned up local distributed task state")
+		}
+
+		for desc, task := range startedTasks {
+			handle, err := provider.StartTask(task)
+			if err != nil {
+				return errors.Wrapf(err, "provider %s start task %v", namespace, desc)
+			}
+
+			s.setRunningTaskHandleWithLock(namespace, desc, handle)
+			s.logger.WithFields(logrus.Fields{
+				"namespace":   namespace,
+				"taskID":      desc.ID,
+				"taskVersion": desc.Version,
+			}).Info("started distributed task execution")
+		}
+
+		s.tasksRunning.
+			WithLabelValues(namespace).
+			Set(float64(len(startedTasks)))
+	}
+
+	enterrors.GoWrapper(s.loop, s.logger)
+
+	return nil
+}
+
+func (s *Scheduler) filterStartedTasks(tasks map[TaskDescriptor]*Task) map[TaskDescriptor]*Task {
+	return filterTasks(tasks, func(task *Task) bool {
+		return task.Status == TaskStatusStarted && !task.FinishedNodes[s.localNode]
+	})
+}
+
+func filterTasks(tasks map[TaskDescriptor]*Task, predicate func(task *Task) bool) map[TaskDescriptor]*Task {
+	filtered := make(map[TaskDescriptor]*Task, len(tasks))
+	for _, task := range tasks {
+		if !predicate(task) {
+			continue
+		}
+
+		filtered[TaskDescriptor{
+			ID:      task.ID,
+			Version: task.Version,
+		}] = task
+	}
+	return filtered
+}
+
+func (s *Scheduler) loop() {
+	ticker := s.clock.NewTicker(s.tickInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.Chan():
+			s.tick()
+		case <-s.stopCh:
+			return
+		}
+	}
+}
+
+func (s *Scheduler) tick() {
+	tasksByNamespace, err := s.listTasks(context.Background())
+	if err != nil {
+		s.sampledLogger.WithSampling(func(l logrus.FieldLogger) {
+			l.WithFields(logrus.Fields{
+				"error": err,
+			}).Error("failed to list distributed tasks")
+		})
+		return
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	for namespace, provider := range s.providers {
+		tasks := tasksByNamespace[namespace]
+
+		// Check that all tasks that are supposed to be running
+		// and launch if they aren't.
+		startedTasks := s.filterStartedTasks(tasks)
+		for _, activeTask := range startedTasks {
+			if _, alreadyLaunched := s.runningTasks[namespace][activeTask.TaskDescriptor]; alreadyLaunched {
+				continue
+			}
+
+			handle, err := provider.StartTask(activeTask)
+			if err != nil {
+				s.sampledLogger.WithSampling(func(l logrus.FieldLogger) {
+					l.WithFields(logrus.Fields{
+						"namespace":   namespace,
+						"taskID":      activeTask.ID,
+						"taskVersion": activeTask.Version,
+						"error":       err,
+					}).Error("failed to start distributed task")
+				})
+				continue
+			}
+
+			s.setRunningTaskHandleWithLock(namespace, activeTask.TaskDescriptor, handle)
+			s.logger.WithFields(logrus.Fields{
```

Comment:
I see that those logged fields are the same throughout this struct wouldn't it be better to have a dedicated method that would always return them? that way you wouldn't have to copy this part of the code everytime you would want to log something
```
func (s *Scheduler) logTaskFields(namespace, activeTask ...) logrus.Fields { ... }
```

Reply:
yeah, good call. Going to refactor